### DISCUSSION TITLE
Request a review on ebean-uuidv7 as Version 7 UUID generator

### DISCUSSION THREAD

**Discussion by rbygrave** [ebean-orm/ebean#3358] [General]
https://github.com/ebean-orm/ebean-uuidv7

This is derived work from https://github.com/cowtowncoder/java-uuid-generator ... but specifically removes the support for monotonically increasing values (as that then requires the generator to be stateful and include locking etc).

That is, this derived work removes that to become stateless and lock free (and minimal, V7 UUID generation only).

Please review and comment etc.

Thanks, Rob.

**Comment by rPraml** [comment]
Hi @rbygrave 

This is a bit difficult to understand:
```
    final byte[] random = new byte[16];
    secureRandom.nextBytes(random);
    final long millis = clock.millis();
    return constructV7((millis << 16) | _toShort(random), _toLong(random));
```

`_toShort` uses byte[0] + byte[1]
`_toLong` calls `_toInt` with offset 2 + 6

so in total, only 10 bytes of 16 are used.

The code is correct, but it was a bit hard to follow the code path (maybe all methods should have an offset parameter?)

Next question: Should we use `new byte[10]` instead of 16?
Could be faster to generate less secure-random data. Could also be slower as we might get misaligned data in the cpu-caches....
which brings me to the next theoretical question, maybe it is better to generate the long from byte0-7 instead of byte2-9 (and use little endian instead of big endian)

Roland

**Reply by rob-bygrave** [reply]
> only 10 bytes of 16 are used

Yes, the code that this has come from supports other UUID versions and so yes this code could be modified to use byte[10] only because there is no intention here to support anything other than Version 7 UUID.

> Should we use new byte[10] instead of 16?

Yes, I think that is a good idea. I don't know about alignment here but I'd be happy to go with byte[10] myself.

> maybe it is better to generate the long from byte0-7

Yes, that sounds like a good idea too. The _toShort, _toLong, _toInt methods can all adjust as necessary for the "This only supports Version 7 UUID" - they've all come from a more generic implementation. Making things simpler here is all good by me.
